workers/unity/Packages/io.improbable.gdk.tools/.CodeGenTemplate/CodeGenerationLib/Jobs/CodegenJob.cs

```diff
@@ -72,27 +77,62 @@ protected void AddContent(string filePath, string fileContents)
             Logger.Trace($"Added generated content for {filePath}.");
         }
 
+        protected void AddJobTarget(string filePath, Func<CodeWriter.CodeWriter> generateFunc)
+        {
+            jobTargets.Add(new JobTarget(Path.Combine(OutputDirectory, filePath), generateFunc));
+        }
+
+        protected void AddJobTarget(string filePath, Func<string> generateFunc)
+        {
+            jobTargets.Add(new JobTarget(Path.Combine(OutputDirectory, filePath), generateFunc));
+        }
+
+        protected void AddGenerators<T>(IEnumerable<T> details, params Func<T, (string relativeFilePath, Func<T, CodeWriter.CodeWriter> generateFunc)>[] generatorJobs) where T : GeneratorInputDetails
+        {
+            jobTargets.AddRange(details.SelectMany(d =>
+            {
+                return generatorJobs.Select(generator =>
+                {
+                    var (filePath, generateImpl) = generator(d);
+                    return new JobTarget(Path.Combine(OutputDirectory, d.NamespacePath, filePath), () => generateImpl(d));
+                });
+            }));
+        }
+
+        protected void AddGenerators<T>(IEnumerable<T> details, params Func<T, (string relativeFilePath, Func<T, string> generateFunc)>[] generatorJobs) where T : GeneratorInputDetails
```

Comment:
Suggestion: Use [delegates](delegates) to effectively "typedef" these closures. This is a beefy type signature 😅 

Note that a delegate != a closure _exactly_, but I think it shouldn't make a difference here. 

[delegates]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/delegates/